src/template/diagnostic.rs

```diff
@@ -0,0 +1,424 @@
+//! The code for error/diagnostics handling is heavily inspiered by
+//! [rust's](https://github.com/rust-lang/rust) compiler. While some code is adpated for use with
+//! punktf, some of it is also a plain copy of it.
+//!
+//! Specifically from those files:
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/lib.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/analyze_source_file.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_parse/src/parser/diagnostics.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/diagnostic.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/diagnostic_builder.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/emitter.rs
+
+use std::borrow::Cow;
+use std::collections::btree_map::Entry;
+use std::collections::{BTreeMap, HashSet};
+
+use color_eyre::owo_colors::OwoColorize;
+
+use super::source::Location;
+use super::span::ByteSpan;
+use crate::template::source::Source;
+
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct DiagnosticSpan {
+	/// A primary span is displayed with `^` bellow the spanned text.
+	pub(super) primary: Vec<ByteSpan>,
+
+	/// A label is displayed as the spanned text together with the label.
+	pub(super) labels: Vec<(ByteSpan, Cow<'static, str>)>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
```

Comment:
Why would we need `Ord` here?

Reply:
Why wouldn't we implement `Ord` for the DiagnosticLevel.
It makes sense to implement as all items in the enum have a total order.
It has no runtime cost and only a small compile time cost as this enum is very small.

Reply:
Because it add buildtime and build size!

Reply:
ye ye ye i agree